### ISSUE TITLE
Should [DisallowNull] apply to value types or nullable value types?

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#36009]
A few options:
1. no warning (as today)
2. warn for passing a possibly null value to `[DisallowNull]int?` parameter
3. disallow `[DisallowNull]` on that parameter. If we do that, then we should probably also disallow on `string?`.

```C#
        [Fact]
        public void DisallowNull_04()
        {
            var source =
@"using System.Runtime.CompilerServices;
class Program
{
    static void F1([DisallowNull]int i) { }
    static void F2([DisallowNull]int? i) { }
    static void M1(int i1)
    {
        F1(i1);
        F2(i1);
    }
    static void M2(int? i2)
    {
        F2(i2); // I expected a warning here (jcouv)
        if (i2 == null) return;
        F2(i2);
    }
}";
            var comp = CreateCompilation(new[] { DisallowNullAttributeDefinition, source }, options: WithNonNullTypesTrue());
            comp.VerifyDiagnostics();
        }
```

Branch `disallow-nullable`

**Comment by CyrusNajmabadi** [comment]
So it's quite possible i haven't thought through this fully, but this just seems wonky to me:

```c#
    static void F2([DisallowNull]int? i) { }
```

Value types already have a great way to indicate they don't allow 'null'.  Simply don't use `Nullable<T>`.  Right?  What's the use case for writing code in this manner?

**Comment by jcouv** [comment]
I think you're right. We could just ignore the attribute (as today), or maybe produce a diagnostic (that attribute is useless when applied to types such as `int?` or `string?`).

**Comment by YairHalberstadt** [comment]
You could have a property which does not accept `null`, but always returns an `int?`.

Or you could be overriding a method which accepts an `int?`, but you will throw on receiving one.

**Comment by YairHalberstadt** [comment]
I don't see how this is different to allowing `[DisallowNull]` on anything that's not a type parameter?

If anything there's more need to do this for value types, since you can't change the nullability of a value type when overriding, but can do so for a reference type.

**Comment by CyrusNajmabadi** [comment]
> If anything there's more need to do this for value types, since you can't change the nullability of a value type when overriding

i feel like this violates the inheritance relationship.  If my base type says it takes an `int?` then it seems broken to me that my subtype woudl not allow it.  it would be similar to the base type stating it takes 'object', but the subtype only takes string.  In other words, this should be contravariant.  But stating `[DisallowNull]` on the subtype breaks that.

**Comment by YairHalberstadt** [comment]
Throwing NotSupportedException in an override is already a common pattern. I don't see how this is different?

**Comment by canton7** [comment]
There's a similar issue somewhere (can't find it) on whether child types should be allowed to tighten the nullability guarantees of their parents (particularly around whether `object.ToString` should return `string` or `string?`).

My recollection is that the BCL would define `virtual string? ToString()`, but derived types could choose to return `string`. I think there was also a discussion on whether `object.Equals` should take an `object` or an `object?`, and whether derived types could override this.

I think it makes sense that the behaviour around `Nullable<T>` and `[DisallowNull]` follows the same rules as nullability annotations on reference types.

**Comment by CyrusNajmabadi** [comment]
> particularly around whether object.ToString should return string or string?).

That's an example of a covariant return type.  Covariance is fine, and you can (and should) absolutely be able to 'narrow' the return type from the broader `string or null` to just `string.  Where this is a problem is in parameters, which need to be contravariant.  There, you have to go 'wider' not 'narrower'.  i.e. if the base says "i only accept strings" the derived can say "but i also support strings or nulls" since that doesn't violate anything.  i.e. anyone seeing the base will only pass strings, which satisfies the actual impl when it goes through the derived side.

**Comment by canton7** [comment]
Sorry, I should have made it clearer that the pertinent part was "I think there was also a discussion on whether object.Equals should take an object or an object?, and whether derived types could override this".

My personal opinion is that I don't hink that inheritance matters here. This is all compile-time. It's not like other sorts of variance where code that was written to expect a particular type can be passed a derived type at runtime. In order for this to matter, the code author has to refactor their code to use a different compile-time type.

That type could be a derived type, yes, but it could also be a base type or a completely unrelated type (with a similar shape). I'm not sure that the rule "if the user refactors their code to replace a type with one of its derived types, there must be no new nullability warnings" is that useful, given that we won't have the rule "if the user refactors their code to replace a type with its base type, there must be no new nullability warnings" or "if the user refactors their code to replace a type with a similar but unrelated type, there must be no new nullability warnings".
